# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (10, 20)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 10 to 20:
the lime is her least liked fruit , but the banana is my least liked .
he saw a old yellow truck .
india is rainy during june , and it is sometimes warm in november .
that cat was my most loved animal .
he dislikes grapefruit , limes , and lemons .
her least liked fruit is the lemon , but his least liked is the grapefruit .
california is never cold during february , but it is sometimes freezing in june .
china is usually pleasant during autumn , and it is usually quiet in october .
paris is never freezing during november , but it is wonderful in october .
the united states is never rainy during january , but it is sometimes mild in october .

French sentences 10 to 20:
la chaux est son moins aimé des fruits , mais la banane est mon moins aimé.
il a vu un vieux camion jaune .
inde est pluvieux en juin , et il est parfois

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_text_elems = source_text.split('\n')
    target_text_elems = target_text.split('\n')
    source_id_text = []
    target_id_text = []
    
    for source_text_elem, target_text_elem in zip(source_text_elems, target_text_elems):
        source_id_text_elem = [source_vocab_to_int[word] for word in source_text_elem.split(' ') if len(word) > 0]
        target_id_text_elem = [target_vocab_to_int[word] for word in target_text_elem.split(' ') if len(word) > 0]
        target_id_text_elem.append(target_vocab_to_int['<EOS>'])
        source_id_text.append(source_id_text_elem)
        target_id_text.append(target_id_text_elem)
        

    print(len(source_id_text))
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

4
Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

137861


# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

print(tf.__version__)
# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

1.1.0
TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    target_sequence_length = tf.placeholder(tf.int32, (None, ), name='target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name='max_target_len')
    
    return (tf.placeholder(tf.int32, (None, None), name='input'), 
            tf.placeholder(tf.int32, (None, None), name='targets'),
            tf.placeholder(tf.float32, name='learning_rate'),
            tf.placeholder(tf.float32, name='keep_prob'),
            target_sequence_length,
            max_target_sequence_length,
            tf.placeholder(tf.int32, (None,), name='source_sequence_length'))
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    go_id = target_vocab_to_int['<GO>']
    pre_proc_target_data = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    go_id_tensor = tf.fill([batch_size, 1], go_id)
    pre_proc_target_data = tf.concat([go_id_tensor, pre_proc_target_data], 1)
    
    return pre_proc_target_data

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)

    def lstm_cell(): 
        return tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1))
    stacked_lstm = tf.contrib.rnn.MultiRNNCell([lstm_cell() for _ in range(num_layers)])
    drop = tf.contrib.rnn.DropoutWrapper(stacked_lstm,
                                         output_keep_prob=keep_prob)
    
    enc_output, enc_state = tf.nn.dynamic_rnn(drop, enc_embed_input, source_sequence_length, dtype=tf.float32)
    
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:
def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    helper = tf.contrib.seq2seq.TrainingHelper(dec_embed_input, target_sequence_length)
    decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, helper, encoder_state, output_layer)
    outputs, _ = tf.contrib.seq2seq.dynamic_decode(decoder, impute_finished=True, maximum_iterations=max_summary_length)
    return outputs
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    start_tokens = tf.fill([batch_size], start_of_sequence_id)
    embeddingHelper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens, end_of_sequence_id)
    decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                              embeddingHelper,
                                              encoder_state,
                                              output_layer=output_layer)
    outputs, _ = tf.contrib.seq2seq.dynamic_decode(decoder,
                   impute_finished=True,
                   maximum_iterations=max_target_sequence_length)
    return outputs

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    #dec_embed_input = tf.contrib.layers.embed_sequence(dec_input, target_vocab_size, decoding_embedding_size)

    def lstm_cell(): 
        return tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1))
    stacked_lstm_dec = tf.contrib.rnn.MultiRNNCell([lstm_cell() for _ in range(num_layers)])
            
    output_fn = Dense(target_vocab_size, kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    #output_fn = lambda x: tf.contrib.layers.fully_connected(x, target_vocab_size, None, scope=decoding_scope)
    
    with tf.variable_scope("decoding"):
        train_logits = decoding_layer_train(encoder_state, # encoder_state
                                      stacked_lstm_dec, # dec_cell
                                      dec_embed_input, # dec_embed_input
                                      target_sequence_length, # target_sequence_length
                                      max_target_sequence_length, # max_target_sequence_length
                                      output_fn, # output_layer
                                      keep_prob) # keep_prob
    with tf.variable_scope("decoding", reuse=True):
        inference_logits = decoding_layer_infer(encoder_state, # encoder_state
                             stacked_lstm_dec, # dec_cell
                             dec_embeddings, # dec_embeddings
                             target_vocab_to_int['<GO>'], # start_of_sequence_id,
                             target_vocab_to_int['<EOS>'], # end_of_sequence_id
                             max_target_sequence_length, # max_target_sequence_length 
                             target_vocab_size, # vocab_size
                             output_fn, # output_layer
                             batch_size, # batch_size
                             keep_prob) # keep_prob
        
    return train_logits, inference_logits

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """

    enc_output, enc_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   enc_embedding_size)
    preproc_data = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    train_logits, inference_logits = decoding_layer(preproc_data, enc_state,
                   target_sequence_length, max_target_sentence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, dec_embedding_size)
    return train_logits, inference_logits

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [14]:
# Number of Epochs
epochs = 32
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 512
# Number of Layers
num_layers = 3
# Embedding Size
encoding_embedding_size = 256
decoding_embedding_size = 256
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.5
display_step = 16

### Build the Graph
Build the graph using the neural network you implemented.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)

Batch and pad the source and target sequences

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths

### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   16/538 - Train Accuracy: 0.4016, Validation Accuracy: 0.4375, Loss: 2.6108
Epoch   0 Batch   32/538 - Train Accuracy: 0.4273, Validation Accuracy: 0.4632, Loss: 2.3384
Epoch   0 Batch   48/538 - Train Accuracy: 0.5050, Validation Accuracy: 0.5172, Loss: 1.9516
Epoch   0 Batch   64/538 - Train Accuracy: 0.4881, Validation Accuracy: 0.5201, Loss: 1.8199
Epoch   0 Batch   80/538 - Train Accuracy: 0.4680, Validation Accuracy: 0.5249, Loss: 1.6578
Epoch   0 Batch   96/538 - Train Accuracy: 0.5329, Validation Accuracy: 0.5398, Loss: 1.4130
Epoch   0 Batch  112/538 - Train Accuracy: 0.4840, Validation Accuracy: 0.5360, Loss: 1.3844
Epoch   0 Batch  128/538 - Train Accuracy: 0.4924, Validation Accuracy: 0.5222, Loss: 1.2229
Epoch   0 Batch  144/538 - Train Accuracy: 0.4629, Validation Accuracy: 0.5028, Loss: 1.3129
Epoch   0 Batch  160/538 - Train Accuracy: 0.4898, Validation Accuracy: 0.5220, Loss: 1.0830
Epoch   0 Batch  176/538 - Train Accuracy: 0.5107, Validation Accuracy

Epoch   2 Batch  384/538 - Train Accuracy: 0.9635, Validation Accuracy: 0.9631, Loss: 0.0257
Epoch   2 Batch  400/538 - Train Accuracy: 0.9695, Validation Accuracy: 0.9412, Loss: 0.0242
Epoch   2 Batch  416/538 - Train Accuracy: 0.9636, Validation Accuracy: 0.9663, Loss: 0.0276
Epoch   2 Batch  432/538 - Train Accuracy: 0.9487, Validation Accuracy: 0.9482, Loss: 0.0280
Epoch   2 Batch  448/538 - Train Accuracy: 0.9689, Validation Accuracy: 0.9501, Loss: 0.0215
Epoch   2 Batch  464/538 - Train Accuracy: 0.9688, Validation Accuracy: 0.9657, Loss: 0.0217
Epoch   2 Batch  480/538 - Train Accuracy: 0.9767, Validation Accuracy: 0.9572, Loss: 0.0206
Epoch   2 Batch  496/538 - Train Accuracy: 0.9805, Validation Accuracy: 0.9549, Loss: 0.0205
Epoch   2 Batch  512/538 - Train Accuracy: 0.9725, Validation Accuracy: 0.9537, Loss: 0.0258
Epoch   2 Batch  528/538 - Train Accuracy: 0.9630, Validation Accuracy: 0.9654, Loss: 0.0257
Epoch   3 Batch   16/538 - Train Accuracy: 0.9762, Validation Accuracy

Epoch   5 Batch  224/538 - Train Accuracy: 0.9848, Validation Accuracy: 0.9538, Loss: 0.0137
Epoch   5 Batch  240/538 - Train Accuracy: 0.9854, Validation Accuracy: 0.9739, Loss: 0.0130
Epoch   5 Batch  256/538 - Train Accuracy: 0.9854, Validation Accuracy: 0.9627, Loss: 0.0092
Epoch   5 Batch  272/538 - Train Accuracy: 0.9943, Validation Accuracy: 0.9574, Loss: 0.0098
Epoch   5 Batch  288/538 - Train Accuracy: 0.9889, Validation Accuracy: 0.9650, Loss: 0.0105
Epoch   5 Batch  304/538 - Train Accuracy: 0.9891, Validation Accuracy: 0.9748, Loss: 0.0122
Epoch   5 Batch  320/538 - Train Accuracy: 0.9957, Validation Accuracy: 0.9824, Loss: 0.0085
Epoch   5 Batch  336/538 - Train Accuracy: 0.9892, Validation Accuracy: 0.9700, Loss: 0.0104
Epoch   5 Batch  352/538 - Train Accuracy: 0.9833, Validation Accuracy: 0.9762, Loss: 0.0201
Epoch   5 Batch  368/538 - Train Accuracy: 0.9966, Validation Accuracy: 0.9743, Loss: 0.0076
Epoch   5 Batch  384/538 - Train Accuracy: 0.9890, Validation Accuracy

Epoch   8 Batch   64/538 - Train Accuracy: 0.9909, Validation Accuracy: 0.9785, Loss: 0.0077
Epoch   8 Batch   80/538 - Train Accuracy: 0.9955, Validation Accuracy: 0.9698, Loss: 0.0052
Epoch   8 Batch   96/538 - Train Accuracy: 0.9903, Validation Accuracy: 0.9817, Loss: 0.0056
Epoch   8 Batch  112/538 - Train Accuracy: 0.9867, Validation Accuracy: 0.9712, Loss: 0.0085
Epoch   8 Batch  128/538 - Train Accuracy: 0.9875, Validation Accuracy: 0.9803, Loss: 0.0109
Epoch   8 Batch  144/538 - Train Accuracy: 0.9895, Validation Accuracy: 0.9712, Loss: 0.0102
Epoch   8 Batch  160/538 - Train Accuracy: 0.9900, Validation Accuracy: 0.9648, Loss: 0.0065
Epoch   8 Batch  176/538 - Train Accuracy: 0.9938, Validation Accuracy: 0.9647, Loss: 0.0067
Epoch   8 Batch  192/538 - Train Accuracy: 0.9972, Validation Accuracy: 0.9739, Loss: 0.0053
Epoch   8 Batch  208/538 - Train Accuracy: 0.9965, Validation Accuracy: 0.9746, Loss: 0.0076
Epoch   8 Batch  224/538 - Train Accuracy: 0.9895, Validation Accuracy

Epoch  10 Batch  432/538 - Train Accuracy: 0.9931, Validation Accuracy: 0.9776, Loss: 0.0063
Epoch  10 Batch  448/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9803, Loss: 0.0033
Epoch  10 Batch  464/538 - Train Accuracy: 0.9918, Validation Accuracy: 0.9783, Loss: 0.0058
Epoch  10 Batch  480/538 - Train Accuracy: 0.9944, Validation Accuracy: 0.9790, Loss: 0.0061
Epoch  10 Batch  496/538 - Train Accuracy: 0.9961, Validation Accuracy: 0.9812, Loss: 0.0051
Epoch  10 Batch  512/538 - Train Accuracy: 0.9963, Validation Accuracy: 0.9748, Loss: 0.0049
Epoch  10 Batch  528/538 - Train Accuracy: 0.9868, Validation Accuracy: 0.9650, Loss: 0.0099
Epoch  11 Batch   16/538 - Train Accuracy: 0.9909, Validation Accuracy: 0.9790, Loss: 0.0058
Epoch  11 Batch   32/538 - Train Accuracy: 0.9959, Validation Accuracy: 0.9790, Loss: 0.0039
Epoch  11 Batch   48/538 - Train Accuracy: 0.9883, Validation Accuracy: 0.9840, Loss: 0.0076
Epoch  11 Batch   64/538 - Train Accuracy: 0.9920, Validation Accuracy

Epoch  13 Batch  272/538 - Train Accuracy: 0.9973, Validation Accuracy: 0.9840, Loss: 0.0032
Epoch  13 Batch  288/538 - Train Accuracy: 0.9932, Validation Accuracy: 0.9696, Loss: 0.0067
Epoch  13 Batch  304/538 - Train Accuracy: 0.9998, Validation Accuracy: 0.9705, Loss: 0.0039
Epoch  13 Batch  320/538 - Train Accuracy: 0.9957, Validation Accuracy: 0.9735, Loss: 0.0045
Epoch  13 Batch  336/538 - Train Accuracy: 0.9976, Validation Accuracy: 0.9718, Loss: 0.0064
Epoch  13 Batch  352/538 - Train Accuracy: 0.9890, Validation Accuracy: 0.9753, Loss: 0.0113
Epoch  13 Batch  368/538 - Train Accuracy: 0.9940, Validation Accuracy: 0.9760, Loss: 0.0047
Epoch  13 Batch  384/538 - Train Accuracy: 0.9914, Validation Accuracy: 0.9661, Loss: 0.0056
Epoch  13 Batch  400/538 - Train Accuracy: 0.9961, Validation Accuracy: 0.9672, Loss: 0.0061
Epoch  13 Batch  416/538 - Train Accuracy: 0.9893, Validation Accuracy: 0.9741, Loss: 0.0057
Epoch  13 Batch  432/538 - Train Accuracy: 0.9991, Validation Accuracy

Epoch  16 Batch  112/538 - Train Accuracy: 0.9992, Validation Accuracy: 0.9750, Loss: 0.0041
Epoch  16 Batch  128/538 - Train Accuracy: 0.9952, Validation Accuracy: 0.9806, Loss: 0.0037
Epoch  16 Batch  144/538 - Train Accuracy: 0.9998, Validation Accuracy: 0.9790, Loss: 0.0038
Epoch  16 Batch  160/538 - Train Accuracy: 0.9948, Validation Accuracy: 0.9755, Loss: 0.0035
Epoch  16 Batch  176/538 - Train Accuracy: 0.9992, Validation Accuracy: 0.9835, Loss: 0.0041
Epoch  16 Batch  192/538 - Train Accuracy: 0.9927, Validation Accuracy: 0.9808, Loss: 0.0062
Epoch  16 Batch  208/538 - Train Accuracy: 0.9998, Validation Accuracy: 0.9838, Loss: 0.0042
Epoch  16 Batch  224/538 - Train Accuracy: 0.9986, Validation Accuracy: 0.9801, Loss: 0.0057
Epoch  16 Batch  240/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9789, Loss: 0.0043
Epoch  16 Batch  256/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9776, Loss: 0.0030
Epoch  16 Batch  272/538 - Train Accuracy: 1.0000, Validation Accuracy

Epoch  18 Batch  480/538 - Train Accuracy: 0.9998, Validation Accuracy: 0.9751, Loss: 0.0019
Epoch  18 Batch  496/538 - Train Accuracy: 0.9980, Validation Accuracy: 0.9863, Loss: 0.0035
Epoch  18 Batch  512/538 - Train Accuracy: 0.9976, Validation Accuracy: 0.9776, Loss: 0.0037
Epoch  18 Batch  528/538 - Train Accuracy: 0.9953, Validation Accuracy: 0.9755, Loss: 0.0058
Epoch  19 Batch   16/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9750, Loss: 0.0022
Epoch  19 Batch   32/538 - Train Accuracy: 0.9976, Validation Accuracy: 0.9767, Loss: 0.0019
Epoch  19 Batch   48/538 - Train Accuracy: 0.9993, Validation Accuracy: 0.9776, Loss: 0.0046
Epoch  19 Batch   64/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9835, Loss: 0.0017
Epoch  19 Batch   80/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9776, Loss: 0.0024
Epoch  19 Batch   96/538 - Train Accuracy: 0.9991, Validation Accuracy: 0.9767, Loss: 0.0033
Epoch  19 Batch  112/538 - Train Accuracy: 1.0000, Validation Accuracy

Epoch  21 Batch  320/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9806, Loss: 0.0019
Epoch  21 Batch  336/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9707, Loss: 0.0022
Epoch  21 Batch  352/538 - Train Accuracy: 0.9967, Validation Accuracy: 0.9730, Loss: 0.0047
Epoch  21 Batch  368/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9817, Loss: 0.0023
Epoch  21 Batch  384/538 - Train Accuracy: 0.9968, Validation Accuracy: 0.9748, Loss: 0.0048
Epoch  21 Batch  400/538 - Train Accuracy: 0.9974, Validation Accuracy: 0.9796, Loss: 0.0015
Epoch  21 Batch  416/538 - Train Accuracy: 0.9998, Validation Accuracy: 0.9751, Loss: 0.0025
Epoch  21 Batch  432/538 - Train Accuracy: 0.9977, Validation Accuracy: 0.9869, Loss: 0.0027
Epoch  21 Batch  448/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9801, Loss: 0.0010
Epoch  21 Batch  464/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9808, Loss: 0.0016
Epoch  21 Batch  480/538 - Train Accuracy: 1.0000, Validation Accuracy

Epoch  24 Batch  160/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9835, Loss: 0.0019
Epoch  24 Batch  176/538 - Train Accuracy: 0.9994, Validation Accuracy: 0.9853, Loss: 0.0010
Epoch  24 Batch  192/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9822, Loss: 0.0010
Epoch  24 Batch  208/538 - Train Accuracy: 0.9998, Validation Accuracy: 0.9792, Loss: 0.0027
Epoch  24 Batch  224/538 - Train Accuracy: 0.9992, Validation Accuracy: 0.9828, Loss: 0.0025
Epoch  24 Batch  240/538 - Train Accuracy: 0.9998, Validation Accuracy: 0.9803, Loss: 0.0022
Epoch  24 Batch  256/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9824, Loss: 0.0013
Epoch  24 Batch  272/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9856, Loss: 0.0023
Epoch  24 Batch  288/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9792, Loss: 0.0020
Epoch  24 Batch  304/538 - Train Accuracy: 0.9973, Validation Accuracy: 0.9753, Loss: 0.0034
Epoch  24 Batch  320/538 - Train Accuracy: 1.0000, Validation Accuracy

Epoch  26 Batch  528/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9842, Loss: 0.0026
Epoch  27 Batch   16/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9824, Loss: 0.0011
Epoch  27 Batch   32/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9833, Loss: 0.0021
Epoch  27 Batch   48/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9856, Loss: 0.0017
Epoch  27 Batch   64/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9918, Loss: 0.0011
Epoch  27 Batch   80/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9860, Loss: 0.0020
Epoch  27 Batch   96/538 - Train Accuracy: 0.9993, Validation Accuracy: 0.9785, Loss: 0.0021
Epoch  27 Batch  112/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9842, Loss: 0.0040
Epoch  27 Batch  128/538 - Train Accuracy: 0.9985, Validation Accuracy: 0.9805, Loss: 0.0037
Epoch  27 Batch  144/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9766, Loss: 0.0022
Epoch  27 Batch  160/538 - Train Accuracy: 1.0000, Validation Accuracy

Epoch  29 Batch  368/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9824, Loss: 0.0013
Epoch  29 Batch  384/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9801, Loss: 0.0020
Epoch  29 Batch  400/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9799, Loss: 0.0011
Epoch  29 Batch  416/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9767, Loss: 0.0005
Epoch  29 Batch  432/538 - Train Accuracy: 0.9982, Validation Accuracy: 0.9806, Loss: 0.0047
Epoch  29 Batch  448/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9771, Loss: 0.0014
Epoch  29 Batch  464/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9860, Loss: 0.0028
Epoch  29 Batch  480/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9767, Loss: 0.0006
Epoch  29 Batch  496/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9716, Loss: 0.0013
Epoch  29 Batch  512/538 - Train Accuracy: 1.0000, Validation Accuracy: 0.9792, Loss: 0.0015
Epoch  29 Batch  528/538 - Train Accuracy: 1.0000, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [20]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    sentence = sentence.lower()
    return [vocab_to_int[word] if word in vocab_to_int else vocab_to_int["<UNK>"] for word in sentence.split()]

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [22]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))

INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [4, 121, 205, 22, 99, 17, 153]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [209, 209, 172, 331, 28, 213, 339, 1]
  French Words: il il un vieux camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.